Fix/tcolor preserve metadata

aviralgarg05 opened the pull request:
# This Pull request: 

## Changes or fixes:

- Add import functools to enable metadata preservation
- Decorate wrapper function with @functools.wraps(original_init) to maintain __name__, __doc__, and __wrapped__ attributes
- Fixes TColor.DefinedColors(1) failure in Jupyter notebooks
- Ensures proper method resolution in introspection-heavy environments

## Checklist:

- [x] tested changes locally
- [x] updated the docs (if necessary)

This PR fixes #20018

aviralgarg05 committed:
[Python] Preserve `__init__` metadata using functools.wraps

This fixes the Jupyter TColor.DefinedColors behavior.

Closes #20018.

dpiparo commented:
Thanks for this code. In order to start the review we would need to also have a test that reproduces the problem. Notebook tests are in this directory https://github.com/root-project/root/tree/master/roottest/python/JupyROOT

aviralgarg05 commented:
I'll go through them and solve

aviralgarg05 committed:
[JupyROOT] Add tests for TColor metadata preservation fix

Add notebook test and unit test to verify that TColor.__init__ metadata
is preserved after pythonization. This ensures that introspection-heavy
environments like Jupyter can properly inspect the method.

  * Add tcolor_definedcolors.ipynb: reproduces original issue #20018
  * Add test_tcolor_metadata.py: unit tests for metadata preservation
  * Update CMakeLists.txt: register notebook test in test suite

This is a regression test for #20018.

guitargeek commented:
This was not the real fix. The actual fix is:
   * https://github.com/root-project/root/pull/22603
